yb-voyager/src/tgtdb/attr_name_registry.go

```diff
@@ -12,45 +12,60 @@ import (
 )
 
 type AttributeNameRegistry struct {
-	dbType    string
-	tdb       TargetDB
-	tconf     *TargetConf
-	attrNames *utils.StructMap[sqlname.NameTuple, []string]
-	mu        sync.Mutex
+	dbType                  string
+	tdb                     TargetDB
+	tconf                   *TargetConf
+	attrNames               *utils.StructMap[sqlname.NameTuple, []string]
+	bestMatchingColumnCache *utils.StructMap[sqlname.NameTuple, map[string]string]
+	mu                      sync.Mutex
 }
 
 func NewAttributeNameRegistry(tdb TargetDB, tconf *TargetConf) *AttributeNameRegistry {
 	return &AttributeNameRegistry{
-		dbType:    tconf.TargetDBType,
-		tdb:       tdb,
-		tconf:     tconf,
-		attrNames: utils.NewStructMap[sqlname.NameTuple, []string](),
+		dbType:                  tconf.TargetDBType,
+		tdb:                     tdb,
+		tconf:                   tconf,
+		attrNames:               utils.NewStructMap[sqlname.NameTuple, []string](),
+		bestMatchingColumnCache: utils.NewStructMap[sqlname.NameTuple, map[string]string](),
 	}
 }
 
 func (reg *AttributeNameRegistry) QuoteAttributeName(tableNameTup sqlname.NameTuple, columnName string) (string, error) {
 	var err error
+	reg.mu.Lock()
+	defer reg.mu.Unlock()
 	targetColumns, ok := reg.attrNames.Get(tableNameTup)
 	if !ok {
-		reg.mu.Lock()
-		// try again in case it's now available
-		targetColumns, ok = reg.attrNames.Get(tableNameTup)
-		if !ok {
-			targetColumns, err = reg.tdb.GetListOfTableAttributes(tableNameTup)
-			log.Infof("columns of table %s in target db: %v", tableNameTup, targetColumns)
-			if err != nil {
-				return "", fmt.Errorf("get list of table attributes: %w", err)
-			}
-			reg.attrNames.Put(tableNameTup, targetColumns)
+		targetColumns, err = reg.tdb.GetListOfTableAttributes(tableNameTup)
+		log.Infof("columns of table %s in target db: %v", tableNameTup, targetColumns)
+		if err != nil {
+			return "", fmt.Errorf("get list of table attributes: %w", err)
 		}
-
-		reg.mu.Unlock()
+		reg.attrNames.Put(tableNameTup, targetColumns)
 	}
-	c, err := reg.findBestMatchingColumnName(columnName, targetColumns)
-	if err != nil {
-		return "", fmt.Errorf("find best matching column name for %q in table %s: %w", columnName, tableNameTup, err)
+	bestMatchingColumnMapForTuple, ok := reg.bestMatchingColumnCache.Get(tableNameTup)
```

Comment:
It would be cleaner to have this logic abstracted away inside the `findBestMatchingColumnName` function. 
That way the API is clean, the caller just calls `findBestMatchingColumnName` . Whether a cache is used internally or not is not the concern of the caller, it's just an implementation detail. 



```
func findBestMatchingColumnName(){
// if present in cache, return. 
// if not, do the actual matching, put it in cache, and return 
}
```

For example, in python, decorators are used to do this quite nicely. https://wiingy.com/learn/python/memoization-using-decorators-in-python/

Reply:
btw, it might make sense to memoize the result of `QuoteAttributeName` func instead of the `findBestMatchingColumnName` . (you've sort of done that, but you could basically check in cache as the very first line of the function; before you even look for all attr names, etc )

Reply:
Yeah, makes sense. In Go, there is nothing similar to decorators I believe but can be implemented with separate wrapper functions. Made the change. Please review it and let me know.